packages/frontend/app/services/channels/vote-verifier.ts

```diff
@@ -0,0 +1,38 @@
+import Service, { inject as service } from '@ember/service';
+import VoteChain from 'emberclear/models/vote-chain';
+
+export default class VoteVerifier extends Service {
+  verify(voteToVerify: VoteChain): boolean {
+    if (voteToVerify.previousVoteChain == undefined) {
+      return true;
+    }
+
+    let voteToVerifyActualString: string = this.generateSortedVoteString(voteToVerify);
+    let voteToVerifyExpectedHash: string = undefined; // NaCl.sign.open(voteToVerify.signature, voteToVerify.key)
+
+    let voteToVerifyActualHash: string = undefined; // Hash voteToVerifyActualString
+    if (voteToVerifyActualHash !== voteToVerifyExpectedHash) {
+      return false;
+    }
+    return this.verify(voteToVerify.previousVoteChain);
+  }
+
+  private generateSortedVoteString(vote: VoteChain): string {
```

Comment:
maybe instead, something with a little less string concatenation would work? like:

```ts
function toSortedPublicKeys(identities) {
  return identities.map(identity => identity.publicKey).sort();
}

let remaining = toSortedPublicKeys(vote.remaining);
let yeses = toSortedPublicKeys(vote.yes);
let nos = toSortedPublicKeys(vote.no);

let payload = [
  remaining,
  yeses,
  nos
  vote.target.publicKey,
  vote.action,
  vote.key.publicKey,
  vote.previousVoteChain.signature,
];

return JSON.stringify(payload);
```

and then you'd need constants for accessing this kind of object -- these are like.. the key.
```ts
REMAINING = 0;
YESES = 1;
NOS = 2;
TARGET_PUBLIC_KEY = 3;
etc
```

and then parsing this would be:

```ts
let payload = JSON.parse(stringifiedPayload);

let yeses = payload[YESES];
```

Reply:
You may even want to initialize the array ahead of time:

```ts
let payload = new Array(8 /* size */);

// ...
payload[YESES] = toSortedPublicKeys(vote.yes);
// ...
```

Reply:
I like the more JSON approach. Fortunately we wouldn't need to be accessing these structures as these will be hashed and compared to the signed value provided to guarantee that the chain hasn't been tampered with.